server/gokbg3/grails-app/services/org/gokb/TippService.groovy

```diff
@@ -0,0 +1,77 @@
+package org.gokb
+
+import com.k_int.ClassUtils
+import com.k_int.ConcurrencyManagerService
+import grails.gorm.transactions.NotTransactional
+import grails.validation.ValidationException
+import net.sf.json.JSON
+import org.gokb.cred.Identifier
+import org.gokb.cred.IdentifierNamespace
+import org.gokb.cred.Imprint
+import org.gokb.cred.KBComponent
+import org.gokb.cred.RefdataCategory
+import org.gokb.cred.ReviewRequest
+import org.gokb.cred.TitleInstance
+import org.gokb.cred.TitleInstancePackagePlatform
+import org.gokb.cred.Package
+import org.gokb.exceptions.MultipleComponentsMatchedException
+import org.grails.web.json.JSONObject
+
+
+class TippService {
+  def componentUpdateService
+  def titleLookupService
+  def messageService
+  def sessionFactory
+  def autoTimestampEventListener
+
+  def copyTitleData(ConcurrencyManagerService.Job job = null) {
+    TitleInstance.withNewSession {
+      autoTimestampEventListener.withoutLastUpdated(TitleInstancePackagePlatform) {
+        int count = 0, index = 0
+        boolean cancelled = false
+        def tippIDs = TitleInstancePackagePlatform.executeQuery('select id from TitleInstancePackagePlatform where status != :status', [status: RefdataCategory.lookup(KBComponent.RD_STATUS, KBComponent.STATUS_DELETED)])
+        log.debug("found ${tippIDs.size()} TIPPs")
+        def tippIDit = tippIDs.iterator()
+        while (tippIDit.hasNext() && !cancelled) {
+          TitleInstancePackagePlatform tipp = TitleInstancePackagePlatform.get(tippIDit.next())
+          index++
+          if (tipp.title) {
+            tipp.title.ids.each { data ->
+              if (['isbn', 'pisbn', 'issn', 'eissn', 'issnl', 'doi', 'zdb', 'isil'].contains(data.namespace.value)) {
+                if (!tipp.ids*.namespace.contains(data.namespace)) {
+                  tipp.ids << data
+                  log.debug("added ID $data in TIPP $tipp")
+                  count++
+                }
+              }
+            }
+            if (!tipp.name || tipp.name == '') {
+              tipp.name = tipp.title.name
+              log.debug("set TIPP name to $tipp.name")
+              count++
+            }
+            tipp.save(flush:true)
+            tipp.finalize()
+          }
+          job?.setProgress(index, tippIDs.size())
+          if (job?.isCancelled()) {
+            cancelled = true
+          }
+          if (count > 100) {
```

Comment:
Ensure this is consistently called, either by incrementing `count` for each TIPP or using smth. like `index % 100 == 0`

Reply:
Autsch! So'n dummer Anfängerfehler. Peinlich, peinlich...

Reply:
The problem I was running in to was that the count is not increased if a TIPP is skipped or nothing is changed. If there are a lot of TIPPs that already have all Identifiers attached to them, those will not increase the counter currently and cause the session to balloon until enough TIPPs are actually changed.

Reply:
I noted that too and changed the implementation to: if (index%100==0)